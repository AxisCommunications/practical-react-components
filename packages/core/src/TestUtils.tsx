import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { PracticalProvider } from './Practical'

/*
  Only used for Jest Snapshot Testing
*/

export const TP = ({
  children,
}: {
  readonly children: React.ReactElement
}): JSX.Element => <PracticalProvider>{children}</PracticalProvider>

/**
 * Mock reference to practical-root
 */
const createNodeMock = (element: React.ReactElement) => {
  if (element.props.id !== 'practical-root') {
    return null
  }

  return {}
}

export const TestRender = (jsx: JSX.Element) => {
  return renderer.create(<TP>{jsx}</TP>, { createNodeMock }).toJSON()
}

export const TestText = (): JSX.Element => <span>Test</span>

/* eslint-disable-next-line react/display-name */
export const TestTextWithForwardRef = React.forwardRef<HTMLElement>(
  ({ ...props }, ref) => (
    <span {...props} ref={ref}>
      Test
    </span>
  )
)
