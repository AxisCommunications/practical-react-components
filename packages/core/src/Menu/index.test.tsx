import React from 'react'
import 'jest-styled-components'

import { Menu, MenuItemProps } from '.'
import { CheckIcon } from 'practical-react-components-icons'
import { TestRender } from '../TestUtils'

const clickHandler = () => console.log('Menu item click')

const items: ReadonlyArray<MenuItemProps> = [
  { icon: CheckIcon, label: 'Apply policy', onClick: clickHandler },
  { label: 'Enter Credentials', onClick: clickHandler },
  { label: 'Pause policy', onClick: clickHandler },
  {
    label: 'Show device report',
    onClick: clickHandler,
    divider: true,
  },
  { label: 'Update firmware', onClick: clickHandler, danger: true },
  { label: 'Restart device', onClick: clickHandler, disabled: true },
]

describe('Menus', () => {
  test('Menu (default)', () => {
    const tree = TestRender(<Menu items={items} />)
    expect(tree).toMatchSnapshot()
  })
  test('Menu (left)', () => {
    const tree = TestRender(<Menu items={items} align="left" />)
    expect(tree).toMatchSnapshot()
  })
  test('Menu (right)', () => {
    const tree = TestRender(<Menu items={items} align="right" />)
    expect(tree).toMatchSnapshot()
  })
  test('Menu (disabled)', () => {
    const tree = TestRender(
      <Menu items={items} align="right" disabled={true} />
    )
    expect(tree).toMatchSnapshot()
  })
})
