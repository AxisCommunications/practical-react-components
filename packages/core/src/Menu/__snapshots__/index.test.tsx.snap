// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menus Menu (default) 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
  overflow: hidden;
  vertical-align: middle;
  height: 24px;
  width: 24px;
}

.c6 > svg {
  fill: currentColor;
  height: 100%;
  width: auto;
}

.c11 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c10 {
  position: absolute;
  z-index: 1;
}

.c2 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  fill: inherit;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c9 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: 50%;
  background-color: transparent;
  outline: none;
  -webkit-transform: scale(0.3);
  -ms-transform: scale(0.3);
  transform: scale(0.3);
  -webkit-transition: -webkit-transform 100ms;
  -webkit-transition: transform 100ms;
  transition: transform 100ms;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  min-width: unset;
  padding: unset;
  outline: none;
  cursor: pointer;
  border: 0;
  color: rgb(102,102,102);
  fill: rgb(102,102,102);
  background-color: transparent;
  -webkit-transition: all 200ms;
  transition: all 200ms;
}

.c3::-moz-focus-inner {
  border: 0;
}

.c3 .c4 {
  border: 2px solid transparent;
}

.c3:hover .c4,
.c3:focus-within .c4 {
  border: 0 solid transparent;
}

.c3:hover .c8,
.c3:focus-within .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:focus:focus .c4 {
  border: 2px solid transparent;
}

.c3:focus:focus .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:active .c8 {
  background-color: rgba(204,204,204,0.24);
  -webkit-transform: scale(1.06);
  -ms-transform: scale(1.06);
  transform: scale(1.06);
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-1"
    >
      <div
        className="c2"
        onBlur={[Function]}
        onKeyDown={[Function]}
      >
        <button
          className="c3"
          disabled={false}
          onClick={[Function]}
          onFocus={[Function]}
          onPointerDown={[Function]}
          onPointerUp={[Function]}
          type="button"
        >
          <div
            className="c4 c5"
          >
            <span
              className="c6 c7 "
              role="img"
              size="medium"
            >
              <svg
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <div
              className="c8 c9"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="c10"
      id="layer-2"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;

exports[`Menus Menu (disabled) 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
  overflow: hidden;
  vertical-align: middle;
  height: 24px;
  width: 24px;
}

.c6 > svg {
  fill: currentColor;
  height: 100%;
  width: auto;
}

.c11 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c10 {
  position: absolute;
  z-index: 1;
}

.c2 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  fill: inherit;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c9 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: 50%;
  background-color: transparent;
  outline: none;
  -webkit-transform: scale(0.3);
  -ms-transform: scale(0.3);
  transform: scale(0.3);
  -webkit-transition: -webkit-transform 100ms;
  -webkit-transition: transform 100ms;
  transition: transform 100ms;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  min-width: unset;
  padding: unset;
  outline: none;
  cursor: pointer;
  border: 0;
  color: rgb(102,102,102);
  fill: rgb(102,102,102);
  background-color: transparent;
  -webkit-transition: all 200ms;
  transition: all 200ms;
  opacity: 0.48;
  pointer-events: none;
  cursor: default;
  box-shadow: none;
}

.c3::-moz-focus-inner {
  border: 0;
}

.c3 .c4 {
  border: 2px solid transparent;
}

.c3:hover .c4,
.c3:focus-within .c4 {
  border: 0 solid transparent;
}

.c3:hover .c8,
.c3:focus-within .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:focus:focus .c4 {
  border: 2px solid transparent;
}

.c3:focus:focus .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:active .c8 {
  background-color: rgba(204,204,204,0.24);
  -webkit-transform: scale(1.06);
  -ms-transform: scale(1.06);
  transform: scale(1.06);
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-7"
    >
      <div
        className="c2"
        onBlur={[Function]}
        onKeyDown={[Function]}
      >
        <button
          className="c3"
          disabled={true}
          onClick={[Function]}
          onFocus={[Function]}
          onPointerDown={[Function]}
          onPointerUp={[Function]}
          type="button"
        >
          <div
            className="c4 c5"
          >
            <span
              className="c6 c7 "
              role="img"
              size="medium"
            >
              <svg
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <div
              className="c8 c9"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="c10"
      id="layer-8"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;

exports[`Menus Menu (left) 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
  overflow: hidden;
  vertical-align: middle;
  height: 24px;
  width: 24px;
}

.c6 > svg {
  fill: currentColor;
  height: 100%;
  width: auto;
}

.c11 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c10 {
  position: absolute;
  z-index: 1;
}

.c2 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  fill: inherit;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c9 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: 50%;
  background-color: transparent;
  outline: none;
  -webkit-transform: scale(0.3);
  -ms-transform: scale(0.3);
  transform: scale(0.3);
  -webkit-transition: -webkit-transform 100ms;
  -webkit-transition: transform 100ms;
  transition: transform 100ms;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  min-width: unset;
  padding: unset;
  outline: none;
  cursor: pointer;
  border: 0;
  color: rgb(102,102,102);
  fill: rgb(102,102,102);
  background-color: transparent;
  -webkit-transition: all 200ms;
  transition: all 200ms;
}

.c3::-moz-focus-inner {
  border: 0;
}

.c3 .c4 {
  border: 2px solid transparent;
}

.c3:hover .c4,
.c3:focus-within .c4 {
  border: 0 solid transparent;
}

.c3:hover .c8,
.c3:focus-within .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:focus:focus .c4 {
  border: 2px solid transparent;
}

.c3:focus:focus .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:active .c8 {
  background-color: rgba(204,204,204,0.24);
  -webkit-transform: scale(1.06);
  -ms-transform: scale(1.06);
  transform: scale(1.06);
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-3"
    >
      <div
        className="c2"
        onBlur={[Function]}
        onKeyDown={[Function]}
      >
        <button
          className="c3"
          disabled={false}
          onClick={[Function]}
          onFocus={[Function]}
          onPointerDown={[Function]}
          onPointerUp={[Function]}
          type="button"
        >
          <div
            className="c4 c5"
          >
            <span
              className="c6 c7 "
              role="img"
              size="medium"
            >
              <svg
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <div
              className="c8 c9"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="c10"
      id="layer-4"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;

exports[`Menus Menu (right) 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
  overflow: hidden;
  vertical-align: middle;
  height: 24px;
  width: 24px;
}

.c6 > svg {
  fill: currentColor;
  height: 100%;
  width: auto;
}

.c11 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c10 {
  position: absolute;
  z-index: 1;
}

.c2 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 {
  fill: inherit;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c9 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: 50%;
  background-color: transparent;
  outline: none;
  -webkit-transform: scale(0.3);
  -ms-transform: scale(0.3);
  transform: scale(0.3);
  -webkit-transition: -webkit-transform 100ms;
  -webkit-transition: transform 100ms;
  transition: transform 100ms;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  min-width: unset;
  padding: unset;
  outline: none;
  cursor: pointer;
  border: 0;
  color: rgb(102,102,102);
  fill: rgb(102,102,102);
  background-color: transparent;
  -webkit-transition: all 200ms;
  transition: all 200ms;
}

.c3::-moz-focus-inner {
  border: 0;
}

.c3 .c4 {
  border: 2px solid transparent;
}

.c3:hover .c4,
.c3:focus-within .c4 {
  border: 0 solid transparent;
}

.c3:hover .c8,
.c3:focus-within .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:focus:focus .c4 {
  border: 2px solid transparent;
}

.c3:focus:focus .c8 {
  background-color: rgba(204,204,204,0.16);
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c3:active .c8 {
  background-color: rgba(204,204,204,0.24);
  -webkit-transform: scale(1.06);
  -ms-transform: scale(1.06);
  transform: scale(1.06);
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-5"
    >
      <div
        className="c2"
        onBlur={[Function]}
        onKeyDown={[Function]}
      >
        <button
          className="c3"
          disabled={false}
          onClick={[Function]}
          onFocus={[Function]}
          onPointerDown={[Function]}
          onPointerUp={[Function]}
          type="button"
        >
          <div
            className="c4 c5"
          >
            <span
              className="c6 c7 "
              role="img"
              size="medium"
            >
              <svg
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <div
              className="c8 c9"
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      className="c10"
      id="layer-6"
    >
      <div
        className="c11"
      />
    </div>
  </div>
</div>
`;
